### ISSUE TITLE
DataParallel multi-gpu results do not match

### ISSUE THREAD

**Issue by taoari** [pytorch/pytorch#4485]
I am using the following scripts for multi-gpu computation using DataParallel:

```
def test():
    
    import torch
    from torchvision import models
    
    batch = torch.autograd.Variable(torch.randn(6,3,224,224))
    
    model = models.resnet18(pretrained=True)
    o1 = model(batch)

    model = model.cuda()
    o2 = model(batch.cuda())

    import os
    os.environ['CUDA_VISIBLE_DEVICES'] = '1,2'
    model = torch.nn.DataParallel(model).cuda()
    o3 = model(batch)
    
    import pdb; pdb.set_trace()
    
if __name__ == '__main__':
    test()
```

The following is the output:

```
(Pdb) torch.__version__
'0.3.0.post4'
(Pdb) o1
Variable containing:
-2.7182 -1.4132 -0.1689  ...  -1.1170  0.9581  2.5264
-0.2388 -1.9526 -1.3678  ...  -1.6317  2.4153  0.3164
 0.4897 -0.4494 -1.2956  ...   1.2032  2.2951 -0.2223
-2.6826 -0.6701 -0.5888  ...  -3.9985  1.1498  3.6315
-0.9542 -1.9719 -1.5599  ...   0.7898 -0.4098 -1.1841
 2.4871  3.4515  1.5563  ...   1.9212  1.0710  0.1708
[torch.FloatTensor of size 6x1000]

(Pdb) o2
Variable containing:
-2.7182 -1.4132 -0.1689  ...  -1.1170  0.9581  2.5264
-0.2388 -1.9526 -1.3678  ...  -1.6317  2.4153  0.3164
 0.4897 -0.4494 -1.2956  ...   1.2032  2.2951 -0.2223
-2.6826 -0.6701 -0.5888  ...  -3.9986  1.1498  3.6315
-0.9542 -1.9719 -1.5599  ...   0.7898 -0.4098 -1.1841
 2.4871  3.4515  1.5563  ...   1.9212  1.0710  0.1708
[torch.cuda.FloatTensor of size 6x1000 (GPU 0)]

(Pdb) o3
Variable containing:
-1.3811 -0.7372  0.4853  ...  -0.5347  1.1808  1.8929
 0.3251 -0.0499 -1.6649  ...  -0.3541  1.3475  0.0513
 1.5659 -0.3755 -0.8420  ...   1.0978  2.3086 -0.6993
-2.5921 -0.5145 -0.2801  ...  -2.1578  0.1405  2.5621
-2.2629 -3.0985 -2.3914  ...  -1.6628  0.5047  0.1904
 1.0781  2.1415  1.3389  ...   0.6119  2.1085  1.6552
[torch.cuda.FloatTensor of size 6x1000 (GPU 0)]

```

The CPU (o1) computation matches the GPU (o2) computation. But why they do not match with multi-GPU (o3) computation?

**Comment by soumith** [comment]
BatchNorm introduces variation in output depending on input batch size.
CPU and single-GPU are running Batch Size of 6
multi-GPU is running batch size of (6 / nGPU)

It's expected that these outputs wont be the same.

**Comment by tommy-qichang** [comment]
I think I face the same issues here. I would say I was shocked that with the same model, same parameters, same input, but just because different GPU number or batch size, I get a different result. 
Is that true that when evaluation, we should use as the same GPU number or as same batch size as the training batch? 
I know it's expected but is there any way to prevent this inconsistency? 
@soumith

**Comment by tommy-qichang** [comment]
Ok, I think I find the answer to this issue: use group norm instead of the batch norm.
[Normalization Alternative to Batch Normalization](https://medium.com/syncedreview/facebook-ai-proposes-group-normalization-alternative-to-batch-normalization-fb0699bffae7)
